src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java

```diff
@@ -0,0 +1,257 @@
+        }
+
+        // The memory file manager stored bytes in the context map, indexed by the class names.
+        return inMemoryClasses.get(name);
+    }
+
+    /**
+     * Load main entry-point class.
+     *
+     * @param parent the class loader to be used as the parent loader
+     * @param mainClassName the fully-qualified name of the application class to load
+     * @return class object representing the desired class
+     * @throws ClassNotFoundException if the class cannot be located
+     * @throws Fault if a modular application class is in the unnamed package
+     */
+    Class<?> loadApplicationClass(ClassLoader parent, String mainClassName) throws ClassNotFoundException, Fault {
+        var moduleInfoBytes = inMemoryClasses.get("module-info");
+        if (moduleInfoBytes == null) {
+            // Trivial case: no compiled module descriptor available, no extra module layer required
+            var memoryClassLoader = new MemoryClassLoader(inMemoryClasses, parent, null, descriptor, this::compileJavaFileByName);
+            return Class.forName(mainClassName, true, memoryClassLoader);
+        }
+
+        // Ensure main class resides in a named package.
+        var lastDotInMainClassName = mainClassName.lastIndexOf('.');
+        if (lastDotInMainClassName == -1) {
+            throw new Fault(Errors.UnnamedPkgNotAllowedNamedModules);
+        }
+
+        var bootLayer = ModuleLayer.boot();
+        var parentLayer = bootLayer;
+        var parentLoader = parent;
+
+        // Optionally create module layer for all modules on the module path.
+        var modulePathFinder = createModuleFinderFromModulePath();
+        var modulePathModules = modulePathFinder.findAll().stream().map(ModuleReference::descriptor).map(ModuleDescriptor::name).toList();
+        if (!modulePathModules.isEmpty()) {
+            var modulePathConfiguration = bootLayer.configuration().resolveAndBind(modulePathFinder, ModuleFinder.of(), Set.copyOf(modulePathModules));
+            var modulePathLayer = ModuleLayer.defineModulesWithOneLoader(modulePathConfiguration, List.of(bootLayer), parent).layer();
+            parentLayer = modulePathLayer;
+            parentLoader = modulePathLayer.findLoader(modulePathModules.getFirst());
+        }
+
+        // Create in-memory module layer for the modular application.
+        var applicationModule = ModuleDescriptor.read(ByteBuffer.wrap(moduleInfoBytes), descriptor::computePackageNames);
+        var memoryFinder = new MemoryModuleFinder(inMemoryClasses, applicationModule, descriptor);
+        var memoryConfig = parentLayer.configuration().resolveAndBind(memoryFinder, ModuleFinder.of(), Set.of(applicationModule.name()));
+        var memoryClassLoader = new MemoryClassLoader(inMemoryClasses, parentLoader, applicationModule, descriptor, this::compileJavaFileByName);
+        var memoryController = ModuleLayer.defineModules(memoryConfig, List.of(parentLayer), __ -> memoryClassLoader);
```

Comment:
It would be helpful if the comment were expanded to explain why memoryClassLoader is needed here. ClassLoaders created by the defineModulesXXX methods are capable to loading modules in memory file systems, the issue is the dynamic compilation (I think).

Reply:
I'll add a comment explaining that the `MemoryClassLoader` is required here to compile referenced types from their source files on-the-fly.

Reply:
Comments added, Alan.

Archie, that's true for the "in-memory" part - but this class loader also tries to find and compile new classes from the default file system on-the-fly, which the class loader in JShell does not (need to do). If both do share enough code, a consolidation would be good to have.